src/content/language/macros.md

```diff
@@ -0,0 +1,287 @@
+  ```
+
+[Add the package]: /guides/packages
+[Add the experiment]: /tools/experiment-flags#using-experiment-flags-with-the-dart-analyzer-command-line-and-ide
+
+### Use the macro
+
+To use the `JsonCodable` macro, append the annotation to the class you want to serialize:
+
+```dart
+import 'package:json/json.dart';
+
+@JsonCodable() // Macro annotation.
+class User {
+ final int? age;
+ final String name;
+ final String username;
+}
+```
+
+The macro introspects the `User` class and derives the implementations of
+`fromJson` and `toJson` using the `User` class's fields.
+
+So, without needing to define them yourself, `toJson` and `fromJson` are now
+available to use on objects of the annotated class:
+
+```dart
+void main() {
+ // Given some arbitrary JSON:
+ var userJson = {
+   'age': 5,
+   'name': 'Roger',
+   'username': 'roger1337'
+ };
+
+ // Use the generated members:
+ var user = User.fromJson(userJson);
+ print(user);
+ print(user.toJson());
+}
+```
+
+### View the generated code
+
+You can optionally view the generated code.
+Click on the "**Go to Augmentation**" link that appears under the annotation
+in your IDE (supported in VSCode and IntelliJ)
+to see how the macro generates `toJson` and `fromJson`.
+
+If you change anything in the annotated class, you can watch the generated augmentation
+adjust in real time alongside your application code:
+
+![A side-by-side gif of the generated augmentation updating as the code it's augmenting is updated](/assets/img/language/macro-augmentation.gif)
+
+### Trigger custom diagnostics
+
+The `JsonCodable` macro has built-in diagnostics that are emmitted just like
+diagnostics from the language itself. For example, if you try to manually
+declare a `toJson` method where the macro is applied, the analyzer will emit
+the error:
+
+```dart
+@JsonCodable()
+class HasToJson {
+ void [!toJson!]() {}
+ // Error: Cannot generate a toJson method due to this existing one.
+}
+```
+
+You can search "`DiagnosticMessage`" in the [the definition of `JsonCodable`][json]
+for other errors the macro will throw. For example, extending a class that isn't
+also serializable, or if field names don't exactly match the key names in the given JSON.
+
+:::note
+To learn more about using the `JsonCodable` macro, like supported field types,
+treatment of null and generics, and more, check out [the README][].
+:::
+
+[the definition of `JsonCodable`]: {{site.repo.dart.sdk}}/blob/master/pkg/json/lib/json.dart
+[the README]: {{site.repo.dart.sdk}}/blob/main/pkg/json/README.md
+
+## The macros language feature
+
+Dart macros are a *static* metaprogramming, or code generation, solution.
+Unlike *runtime* code generation solutions (like [build_runner][]),
+macros are fully integrated into the Dart language and executed by the compiler.
+This makes macros much more efficient than relying on an secondary tool:
+
+- **Nothing extra to run**;
+ simply call `dart ` / `flutter run <your app>` and the macro builds with your code.
```

Comment:
would avoid this bullet which suggests that macros aren't expanded until you run

Reply:
Good point, how about changing the wording to:

```
 macros build in real-time as you write your code.
```

Reply:
yeah, that works